cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java

```diff
@@ -74,88 +83,160 @@
-    this.type = metadataEntity.type;
+  static {
+    Map<String, String[][]> typesToKeys = new HashMap<>();
+    typesToKeys.put(NAMESPACE, new String[][]{{NAMESPACE}});
+    typesToKeys.put(DATASET, new String[][]{{NAMESPACE, DATASET}, {DATASET}});
+    typesToKeys.put(STREAM, new String[][]{{NAMESPACE, STREAM}});
+    typesToKeys.put(APPLICATION, new String[][]{{NAMESPACE, APPLICATION, VERSION}, {NAMESPACE, APPLICATION}});
+    typesToKeys.put(ARTIFACT, new String[][]{{NAMESPACE, ARTIFACT, VERSION}});
+    typesToKeys.put(VIEW, new String[][]{{NAMESPACE, STREAM, VIEW}});
+    typesToKeys.put(PROGRAM, new String[][]{{NAMESPACE, APPLICATION, VERSION, TYPE, PROGRAM},
+      {NAMESPACE, APPLICATION, TYPE, PROGRAM}});
+    typesToKeys.put(SCHEDULE, new String[][]{{NAMESPACE, APPLICATION, VERSION, SCHEDULE},
+      {NAMESPACE, APPLICATION, SCHEDULE}});
+    typesToKeys.put(FLOWLET, new String[][]{{NAMESPACE, APPLICATION, VERSION, FLOW, FLOWLET},
+      {NAMESPACE, APPLICATION, FLOW, FLOWLET}});
+    typesToKeys.put(PROGRAM_RUN, new String[][]{{NAMESPACE, APPLICATION, VERSION, TYPE, PROGRAM, PROGRAM_RUN},
+      {NAMESPACE, APPLICATION, TYPE, PROGRAM, PROGRAM_RUN}});
+    TYPES_TO_KEY_SEQUENCES = Collections.unmodifiableMap(typesToKeys);
   }
 
   /**
-   * Creates a {@link MetadataEntity} representing the given datasetName. To create a {@link MetadataEntity} for a
-   * dataset in a specified namespace please use {@link MetadataEntity#ofDataset(String, String)}.
-   *
-   * @param datasetName the name of the dataset
-   * @return {@link MetadataEntity} representing the dataset name
+   * Used by Builder to build the {@link MetadataEntity}
    */
-  public static MetadataEntity ofDataset(String datasetName) {
-    return new MetadataEntity(DATASET, datasetName);
+  private MetadataEntity(Map<String, String> details, String type) {
+    this.details = details;
+    this.type = type;
   }
 
   /**
-   * Creates a {@link MetadataEntity} representing the given datasetName in the specified namespace.
-   *
-   * @param namespace the name of the namespace
-   * @param datasetName the name of the dataset
-   * @return {@link MetadataEntity} representing the dataset name
+   * Builder for {@link MetadataEntity}
    */
-  public static MetadataEntity ofDataset(String namespace, String datasetName) {
-    return MetadataEntity.ofNamespace(namespace).appendAsType(MetadataEntity.DATASET, datasetName);
+  public static class Builder {
+    private final LinkedHashMap<String, String> parts;
+    private String type;
+
+    Builder() {
+      parts = new LinkedHashMap<>();
+    }
+
+    private Builder(MetadataEntity metadataEntity) {
+      this.parts = new LinkedHashMap<>(metadataEntity.details);
+      this.type = metadataEntity.type;
+    }
+
+    /**
+     * Put the given key with the given value in {@link MetadataEntity.Builder}.
+     * The returned {@link MetadataEntity.Builder} is of the same type. If the type needs to
```

Comment:
a little confusing. Isn't it the other way round? By default, the last key appended becomes the type - unless appendAsType has been called.

Reply:
This one got fixed in addressing other comment. Yes, last key will not be type if type is not